smelter-core/src/pipeline/utils/input_buffer.rs

```diff
@@ -0,0 +1,198 @@
+        if next_pts > self.sync_point.elapsed() + self.max_buffer {
+            let first_pts = self.state.set_over_max_buffer(next_pts);
+            if next_pts.saturating_sub(first_pts) > STABLE_STATE_DURATION {
+                self.dynamic_buffer = self.dynamic_buffer.saturating_sub(DECREMENT_DURATION);
+            }
+        } else if next_pts > self.sync_point.elapsed() + self.desired_buffer {
+            self.state.set_ok();
+        } else if next_pts > self.sync_point.elapsed() + self.min_buffer {
+            trace!(
+                old=?self.dynamic_buffer,
+                new=?self.dynamic_buffer + INCREMENT_DURATION,
+                "Increase latency optimized buffer"
+            );
+            self.state.set_to_small();
+            self.dynamic_buffer += INCREMENT_DURATION;
+        } else {
+            let new_buffer = (self.sync_point.elapsed() + self.desired_buffer).saturating_sub(pts);
+            debug!(
+                old=?self.dynamic_buffer,
+                new=?new_buffer,
+                "Increase latency optimized buffer (force)"
+            );
+            self.state.set_to_small();
+            // adjust buffer so:
+            // pts + self.dynamic_buffer == self.sync_point.elapsed() + self.desired_buffer
+            self.dynamic_buffer = new_buffer
+        }
+
+        pts + self.dynamic_buffer
+    }
+}
+
+enum LatencyOptimizedBufferState {
+    Ok,
+    ToSmallBuffer,
+    OverMaxBuffer { first_pts: Duration },
+}
+
+impl LatencyOptimizedBufferState {
+    fn set_over_max_buffer(&mut self, pts: Duration) -> Duration {
+        match &self {
+            LatencyOptimizedBufferState::OverMaxBuffer { first_pts } => *first_pts,
+            _ => {
+                *self = LatencyOptimizedBufferState::OverMaxBuffer { first_pts: pts };
+                pts
+            }
+        }
+    }
+
+    fn set_to_small(&mut self) {
+        *self = LatencyOptimizedBufferState::ToSmallBuffer
+    }
+
+    fn set_ok(&mut self) {
+        *self = LatencyOptimizedBufferState::Ok
+    }
+}
+
+pub(crate) struct AdaptiveBuffer {
+    sync_point: Instant,
+    desired_buffer: Duration,
+    dynamic_buffer: Duration,
+    min_buffer: Duration,
+}
+
+impl AdaptiveBuffer {
+    fn new(ctx: &PipelineCtx) -> Self {
+        Self {
+            sync_point: ctx.queue_sync_point,
+            desired_buffer: ctx.default_buffer_duration,
+            min_buffer: Duration::min(Duration::from_millis(20), ctx.default_buffer_duration),
+            dynamic_buffer: ctx.default_buffer_duration,
+        }
+    }
+
+    fn pts_with_buffer(&mut self, pts: Duration) -> Duration {
+        const INCREMENT_DURATION: Duration = Duration::from_micros(100);
+
+        let next_pts = pts + self.dynamic_buffer;
+        if next_pts > self.sync_point.elapsed() + self.desired_buffer {
+            // ok
+        } else if next_pts > self.sync_point.elapsed() + self.min_buffer {
+            debug!(
```

Comment:
this will hit only if current PTS is between 20ms and 80ms ahead of a queue. For protocols like HLS it will never hit